abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+the in-VM memory allocator.
+
+Plugin must return `memory_data` pointing to the start of the allocated
+memory.
+
+Returning `0` indicates failure.
+
+
+## Context lifecycle
+
+### `proxy_on_context_create`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) parent_context_id`
+* returns:
+  - none
+
+Called when the host creates a new context (`context_id`).
+
+When `parent_context_id` is `0` then a new plugin ("root") context
+is created, otherwise a new per-stream context is created.
+
+
+### `proxy_on_done`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (bool) completed`
+
+Called when the host is done processing context (`context_id`).
+
+Plugin must return one of the following values:
+- `true` to allow the host to finalize and delete context.
+- `false` to indicate that the context is still being used,
+  and that plugin is going to call [`proxy_done`] later to
+  allow the host to finalize and delete that context.
+
+
+### `proxy_on_log`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called after the host is done processing context, but before
+releasing its state.
+
+This callback can be used for generating final log entries.
+
+
+### `proxy_on_delete`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - none
+
+Called when the host removes the context (`context_id`).
+
+This is used to signal that the plugin should stop tracking active
+context and remove all associated state.
+
+
+### `proxy_done`
+
+* params:
+  - none
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Indicates to the host that the plugin is done processing active
+context.
+
+This should be used after returning `false` in [`proxy_on_done`].
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when active context was not pending finalization.
+
+
+### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Changes the effective context to `context_id`.
+
+This can be used to change active context, e.g. during
+[`proxy_on_http_call_response`], [`proxy_on_grpc_receive`]
+and/or [`proxy_on_queue_ready`] callbacks.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `context_id`.
+
+
+## Configuration
+
+### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the WebAssembly Virtual Machine.
+
+Its configuration (of `vm_configuration_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to `VM_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed.
+
+
+### `proxy_on_configure`
+
+* params:
+  - `i32 (uint32_t) context_id`
```

Comment:
Thinking about this more, maybe that would be overly verbose. But at the same time it could be helpful to designate which callbacks carry context_ids for root vs. non-root contexts.

I'm thinking in particular of callbacks like `proxy_on_http_call_response`, `proxy_on_grpc_receive`, etc., where IIRC the context_id is for the root context rather than the stream context that initiated the call. It might be good to indicate this somehow, either by using the name `root_context_id` vs. `context_id` to indicate which is passed, or by adding some text about callbacks that are delivered to the root context rather than stream context. WDYT?